### PR TITLE
[DRAFT] Support add training

### PR DESCRIPTION
**Pull request by Aeren1564**
Draft PR for supporting Add training

### CODE REVIEW

File: runtime/onert/backend/train/ops/BinaryArithmeticLayer.cc [hunk 1]
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2023 Samsung Electronics Co., Ltd. All Rights Reserved
```
Comment (jyoungyun) [inline runtime/onert/backend/train/ops/BinaryArithmeticLayer.cc:2]:
```suggestion
 * Copyright (c) 2024 Samsung Electronics Co., Ltd. All Rights Reserved
```

File: runtime/onert/backend/train/ops/BinaryArithmeticLayer.cc [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                                      IPortableTensor *back_prop_rhs,
+                                      const IPortableTensor *back_prop_output,
+                                      const ArithmeticType arithmetic_type,
+                                      const ir::Activation activation)
+{
+  cpu::ops::BinaryArithmeticLayer::configure(lhs, rhs, output, arithmetic_type, activation);
+
+  _back_prop_lhs = back_prop_lhs;
+  _back_prop_rhs = back_prop_rhs;
+  _back_prop_output = back_prop_output;
```
Comment (jyoungyun) [inline runtime/onert/backend/train/ops/BinaryArithmeticLayer.cc:51]: Please throw an error if the `artimetic_type` is not `add`. :)

File: runtime/onert/backend/train/ops/BinaryArithmeticLayer.cc (hunk 1 above)
Comment (Aeren1564) [reply runtime/onert/backend/train/ops/BinaryArithmeticLayer.cc]: Fixed.

File: runtime/onert/backend/train/ops/BinaryArithmeticLayer.cc (hunk 2 above)
Comment (Aeren1564) [reply runtime/onert/backend/train/ops/BinaryArithmeticLayer.cc:55]: It now throws an error if op is not add.

File: compute/cker/include/cker/train/operation/BinaryArithmetic.h [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+		}
+		break;
+
+		default:
+			throw std::runtime_error{"Unsupported Binary Arithmetic Operation"};
+	}
+}
+
+template <typename T, >
+void BinaryArithmeticRHSGrad(const Shape &lhs_shape, const T *lhs_data, const Shape &rhs_shape, const Shape &rhs_data, const Shape &incoming_shape, const T *incoming_data, const Shape &grad_shape,
```
[the remark this answers is not in the document]
Comment (Aeren1564) [reply compute/cker/include/cker/train/operation/BinaryArithmetic.h]: I guess there's no need to split them. They have been merged.

File: runtime/onert/backend/train/ops/BinaryArithmeticLayer.cc [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    case ir::Activation::RELU:
+      nnfw::cker::train::ReLUGrad(getShape(_output), getBuffer<float>(_output),
+                                  getShape(_back_prop_output), getBuffer<float>(_back_prop_output),
+                                  getShape(_act_back_prop_output.get()),
+                                  getBuffer<float>(_act_back_prop_output.get()));
+      backprop_act = _act_back_prop_output.get();
+      break;
+    default:
+      throw std::runtime_error("train ConvolutionLayer: Unsupported activation type yet");
+  }
```
Comment (zetwhite) [inline runtime/onert/backend/train/ops/BinaryArithmeticLayer.cc:85]: just FYI. 

I recently updated this related part. (related PR : #12492 )
So, instead of directly calling activation grad(), you could use a function in OperationUtils.h

https://github.com/Samsung/ONE/blob/f444dab7e2de17b7cbcdff39e2d1bc647e3103bc/runtime/onert/backend/train/ops/OperationUtils.h#L39-L62
 
like here : 

https://github.com/Samsung/ONE/blob/f444dab7e2de17b7cbcdff39e2d1bc647e3103bc/runtime/onert/backend/train/ops/ConvolutionLayer.cc#L136-L146

File: compute/cker/include/cker/train/operation/BinaryArithmetic.h [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+{
+  BroadcastTo(incoming_shape, const_cast<T *>(incoming_data), lhs_grad_shape, lhs_grad_data);
+  BroadcastTo(incoming_shape, const_cast<T *>(incoming_data), rhs_grad_shape, rhs_grad_data);
+  switch (arithmetic_type)
+  {
+    case ArithmeticType::kAdd:
+      break;
+    case ArithmeticType::kSub:
+    case ArithmeticType::kMul:
+    case ArithmeticType::kDiv:
```
Comment (jyoungyun) [inline compute/cker/include/cker/train/operation/BinaryArithmetic.h:52]: Are other operations always required to broadcast?

File: runtime/onert/backend/train/ops/BinaryArithmeticLayer.cc (hunk 4 above)
Comment (Aeren1564) [reply runtime/onert/backend/train/ops/BinaryArithmeticLayer.cc]: Thanks for letting me know :)
Updated.

**Comment by Aeren1564** [comment]
I managed to code something with add layer that seems to work.
I'll create next PR after I figure out how to do this with ONERT and confirm that the implementation is correct.

```python
import numpy as np
import tensorflow as tf
import keras
from tensorflow.keras import optimizers
from keras import layers
import tensorflow_datasets as tfds

input_lhs = tf.keras.layers.Input(shape=(10))
input_rhs = tf.keras.layers.Input(shape=(10))
lhs = tf.keras.layers.Dense(10)(input_lhs)
rhs = tf.keras.layers.Dense(10)(input_rhs)
res_add = tf.keras.layers.Add()([lhs, rhs])
output = tf.keras.layers.Dense(10)(res_add)
model = tf.keras.models.Model(inputs=[input_lhs, input_rhs], outputs=output, name="add_training")

model.summary()

model.compile(
	optimizer='adam',
	loss="mean_squared_error",
	metrics=['mae'],
)

np.random.seed(123)

data_lhs = np.random.rand(3000, 10) * 100
data_rhs = np.random.rand(3000, 10) * 100
coef_lhs, coef_rhs = np.random.rand(10, 10), np.random.rand(10, 10)
data_res = np.array([(np.matmul(coef_lhs, x[0]) + np.matmul(coef_rhs, x[1])) for x in zip(data_lhs, data_rhs)], dtype=float)

model.fit(
	[data_lhs, data_rhs],
	data_res,
	epochs=100,
	batch_size=32,
	verbose=2
)
```

File: compute/cker/include/cker/train/operation/BinaryArithmetic.h (hunk 5 above)
Comment (Aeren1564) [reply compute/cker/include/cker/train/operation/BinaryArithmetic.h:52]: I was planning to work with broadcasted data (which, to my understanding, is simple copying) for other OPs.
Is there a better alternative?

**Comment by Aeren1564** [comment]
> > after I figure out how to do this with ONERT
> 
> I uploaded my local test environment [here](https://github.sec.samsung.net/sseung-youn/step-by-step/tree/main/source/circle%2B). Although code is not organized well, You could refer my env. (if necessary)

Thanks :)

File: compute/cker/include/cker/train/operation/BinaryArithmetic.h (hunk 5 above)
Comment (jyoungyun) [reply compute/cker/include/cker/train/operation/BinaryArithmetic.h:52]: Well, if there is a way to calculate both broadcast and gradient at the same time, it would be great. But it is not related to this PR, so let's think about it later. :)

**Comment by jyoungyun** [comment]
> I managed to code something with add layer that seems to work.

This code is for creating and training model. After doing this, you need to save the trained model to both TensorFlow and Circle model. And then, you have to load each model file and train the model using the same dataset. :)

**Comment by Aeren1564** [comment]
Tensorflow
---
```asm
__________________________________________________________________________________________________
Epoch 1/5
150/150 [==============================] - 0s 587us/step - loss: 553.6255 - mae: 18.5547
Epoch 2/5
150/150 [==============================] - 0s 641us/step - loss: 387.7979 - mae: 15.4453
Epoch 3/5
150/150 [==============================] - 0s 628us/step - loss: 274.3452 - mae: 12.8840
Epoch 4/5
150/150 [==============================] - 0s 574us/step - loss: 190.2583 - mae: 10.5831
Epoch 5/5
150/150 [==============================] - 0s 573us/step - loss: 132.0477 - mae: 8.6221
```


ONERT-Train
---
```asm
/home/aeren/Repos/ONE/Product/x86_64-linux.debug/out/bin/onert_train --modelfile /home/aeren/Repos/Scripts/_Product/circle+/result/model20240126_1032/model.circle --load_input:raw /home/aeren/Repos/Scripts/_Product/circle+/data/input.bin --load_expected:raw /home/aeren/Repos/Scripts/_Product/circle+/data/res.bin --epoch 5 --batch_size 20 --learning_rate 0.001 --loss 1 --loss_reduction_type 1 --optimizer 2 
Model Expected Filename /home/aeren/Repos/Scripts/_Product/circle+/data/res.bin
Model Input Filename /home/aeren/Repos/Scripts/_Product/circle+/data/input.bin
Model Filename /home/aeren/Repos/Scripts/_Product/circle+/result/model20240126_1032/model.circle
== training parameter ==
- learning_rate   = 0.001
- batch_size      = 20
- loss_info       = {loss = mean squared error, reduction = sum over batch size}
- optimizer       = adam
========================
Epoch 1/5 - time: 0.319ms/step - loss: [0] 553.6250
Epoch 2/5 - time: 0.305ms/step - loss: [0] 387.7969
Epoch 3/5 - time: 0.289ms/step - loss: [0] 274.3438
Epoch 4/5 - time: 0.302ms/step - loss: [0] 190.2571
Epoch 5/5 - time: 0.291ms/step - loss: [0] 132.0467
```

The loss values are identical :)

**Comment by Aeren1564** [comment]
I slightly modified @zetwhite's code and ran training on the model below, with identical input/output for Tensorflow and ONERT.

Model with Subtract
---
```python
    input_lhs = tf.keras.layers.Input(shape=(10))
    input_rhs = tf.keras.layers.Input(shape=(10))
    lhs = tf.keras.layers.Dense(10)(input_lhs)
    rhs = tf.keras.layers.Dense(10)(input_rhs)
    res_sub = tf.keras.layers.subtract([lhs, rhs])
    output = tf.keras.layers.Dense(10)(res_sub)
    model = tf.keras.models.Model(inputs=[input_lhs, input_rhs], outputs=output, name="subtract_training")
```

Data
---
```python
    np.random.seed(123)
    data_lhs = np.random.rand(3000, 10).astype(np.float32) * 100
    data_rhs = np.random.rand(3000, 10).astype(np.float32) * 100
    coef_lhs, coef_rhs = np.random.rand(10, 10).astype(np.float32), np.random.rand(10, 10).astype(np.float32)
    data_res = np.array([(np.matmul(coef_lhs, x[0]) + np.matmul(coef_rhs, x[1])) for x in zip(data_lhs, data_rhs)], dtype=np.float32)
```

Tensorflow
---
```asm
__________________________________________________________________________________________________
Epoch 1/5
150/150 [==============================] - 0s 578us/step - loss: 647.6946 - mae: 20.2884
Epoch 2/5
150/150 [==============================] - 0s 585us/step - loss: 484.7763 - mae: 17.5722
Epoch 3/5
150/150 [==============================] - 0s 581us/step - loss: 382.2013 - mae: 15.5862
Epoch 4/5
150/150 [==============================] - 0s 592us/step - loss: 305.6406 - mae: 13.9046
Epoch 5/5
150/150 [==============================] - 0s 629us/step - loss: 241.3343 - mae: 12.2882
```


ONERT-Train
---
```asm
/home/aeren/Repos/ONE/Product/x86_64-linux.debug/out/bin/onert_train --modelfile /home/aeren/Repos/Scripts/_Product/circle+/result/model20240126_1244/model.circle --load_input:raw /home/aeren/Repos/Scripts/_Product/circle+/data/input.bin --load_expected:raw /home/aeren/Repos/Scripts/_Product/circle+/data/res.bin --epoch 5 --batch_size 20 --learning_rate 0.001 --loss 1 --loss_reduction_type 1 --optimizer 2 
Model Expected Filename /home/aeren/Repos/Scripts/_Product/circle+/data/res.bin
Model Input Filename /home/aeren/Repos/Scripts/_Product/circle+/data/input.bin
Model Filename /home/aeren/Repos/Scripts/_Product/circle+/result/model20240126_1244/model.circle
== training parameter ==
- learning_rate   = 0.001
- batch_size      = 20
- loss_info       = {loss = mean squared error, reduction = sum over batch size}
- optimizer       = adam
========================
Epoch 1/5 - time: 0.327ms/step - loss: [0] 647.6941
Epoch 2/5 - time: 0.318ms/step - loss: [0] 484.7753
Epoch 3/5 - time: 0.302ms/step - loss: [0] 382.2000
Epoch 4/5 - time: 0.317ms/step - loss: [0] 305.6393
Epoch 5/5 - time: 0.303ms/step - loss: [0] 241.3329
```

The loss values are (almost) identical :)

**Comment by Aeren1564** [comment]
Model with Multiply
---
```python
    input_lhs = tf.keras.layers.Input(shape=(10))
    input_rhs = tf.keras.layers.Input(shape=(10))
    lhs = tf.keras.layers.Dense(10)(input_lhs)
    rhs = tf.keras.layers.Dense(10)(input_rhs)
    res_mul = tf.keras.layers.Multiply()([lhs, rhs])
    output = tf.keras.layers.Dense(10)(res_mul)
    model = tf.keras.models.Model(inputs=[input_lhs, input_rhs], outputs=output, name="multiply_training")
```

Data
---
```python
    np.random.seed(123)
    data_lhs = np.random.rand(3000, 10).astype(np.float32) * 100
    data_rhs = np.random.rand(3000, 10).astype(np.float32) * 100
    coef_lhs, coef_rhs = np.random.rand(10, 10).astype(np.float32), np.random.rand(10, 10).astype(np.float32)
    data_res = np.array([(np.matmul(coef_lhs, x[0]) + np.matmul(coef_rhs, x[1])) for x in zip(data_lhs, data_rhs)], dtype=np.float32)
```

Tensorflow
---
```asm
__________________________________________________________________________________________________
Epoch 1/5
150/150 [==============================] - 0s 688us/step - loss: 6594.2104 - mae: 64.8041
Epoch 2/5
150/150 [==============================] - 0s 676us/step - loss: 5605.8013 - mae: 59.9134
Epoch 3/5
150/150 [==============================] - 0s 573us/step - loss: 5306.5811 - mae: 58.3457
Epoch 4/5
150/150 [==============================] - 0s 548us/step - loss: 5146.8296 - mae: 57.5273
Epoch 5/5
150/150 [==============================] - 0s 547us/step - loss: 5031.8623 - mae: 56.9231
```


ONERT-Train
---
```asm
/home/aeren/Repos/ONE/Product/x86_64-linux.debug/out/bin/onert_train --modelfile /home/aeren/Repos/Scripts/_Product/circle+/result/model20240126_1329/model.circle --load_input:raw /home/aeren/Repos/Scripts/_Product/circle+/data/input.bin --load_expected:raw /home/aeren/Repos/Scripts/_Product/circle+/data/res.bin --epoch 5 --batch_size 20 --learning_rate 0.001 --loss 1 --loss_reduction_type 1 --optimizer 2 
Model Expected Filename /home/aeren/Repos/Scripts/_Product/circle+/data/res.bin
Model Input Filename /home/aeren/Repos/Scripts/_Product/circle+/data/input.bin
Model Filename /home/aeren/Repos/Scripts/_Product/circle+/result/model20240126_1329/model.circle
== training parameter ==
- learning_rate   = 0.001
- batch_size      = 20
- loss_info       = {loss = mean squared error, reduction = sum over batch size}
- optimizer       = adam
========================
Epoch 1/5 - time: 0.300ms/step - loss: [0] 6594.2061
Epoch 2/5 - time: 0.288ms/step - loss: [0] 5605.7979
Epoch 3/5 - time: 0.289ms/step - loss: [0] 5306.5791
Epoch 4/5 - time: 0.291ms/step - loss: [0] 5146.8281
Epoch 5/5 - time: 0.291ms/step - loss: [0] 5031.8633
```

File: compute/cker/include/cker/train/operation/BinaryArithmetic.h [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    {
+      auto const in_map = MapAsMatrixWithLastDimAsRows(incoming_data, incoming_shape);
+      auto const lhs_map = MapAsMatrixWithLastDimAsRows(lhs_data, lhs_shape);
+      auto const rhs_map = MapAsMatrixWithLastDimAsRows(rhs_data, rhs_shape);
+      auto lhs_grad_map = MapAsMatrixWithLastDimAsRows(lhs_grad_data, lhs_grad_shape);
+      auto rhs_grad_map = MapAsMatrixWithLastDimAsRows(rhs_grad_data, rhs_grad_shape);
+
+      lhs_grad_map = in_map.array() / rhs_map.array();
+      rhs_grad_map = in_map.array() * -lhs_map.array() / rhs_map.array() / rhs_map.array();
```
Comment (Aeren1564) [inline compute/cker/include/cker/train/operation/BinaryArithmetic.h:86]: I'm seeing weird outputs :/
Is there something wrong with the following?

$L$: LHS
$R$: RHS
$O$: Output (of division)
$X$: Output (of entire model)

$L / R = O$ (Elementwise division)

$\frac{\partial X}{\partial L} = \frac{\partial O}{\partial L} \cdot \frac{\partial X}{\partial O} = \frac{1}{R} \cdot \frac{\partial X}{\partial O}$
$\frac{\partial X}{\partial R} = \frac{\partial O}{\partial R} \cdot \frac{\partial X}{\partial O} = -\frac{L}{R^2} \cdot \frac{\partial X}{\partial O}$

**Comment by Aeren1564** [comment]
@nnfw-bot test tizen-gbs

**Comment by Aeren1564** [comment]
@nnfw-bot test onert-cross-debug

**Comment by Aeren1564** [comment]
@nnfw-bot test onert-cross-release

**Comment by Aeren1564** [comment]
TODO: try other optimizers for division

**Comment by Aeren1564** [comment]
I've tried following optimizers for division but all of them showed loss values differing from that of tensorflow :/

```
SGD, RMSProp, Adam, Adadelta
```

**Comment by Aeren1564** [comment]
@nnfw-bot test onert-cross-debug

**Comment by Aeren1564** [comment]
@nnfw-bot test onert-cross-release
